src/contracts/GPv2Settlement.sol

```diff
@@ -140,6 +142,69 @@ contract GPv2Settlement is GPv2Signing, ReentrancyGuard, StorageAccessible {
         emit Settlement(msg.sender);
     }
 
+    /// @dev Settle an order directly against Balancer V2 pools.
+    function swap(
+        IVault.SwapRequest[] calldata swaps,
+        IERC20[] calldata tokens,
+        GPv2Trade.Data calldata trade
+    ) external nonReentrant onlySolver {
+        RecoveredOrder memory recoveredOrder = allocateRecoveredOrder();
+        GPv2Order.Data memory order = recoveredOrder.data;
+        recoverOrderFromTrade(recoveredOrder, tokens, trade);
+
+        IVault.SwapKind kind =
+            order.kind == GPv2Order.SELL
+                ? IVault.SwapKind.GIVEN_IN
+                : IVault.SwapKind.GIVEN_OUT;
+
+        IVault.FundManagement memory funds;
+        funds.sender = recoveredOrder.owner;
+        funds.fromInternalBalance = order.useInternalSellTokenBalance;
+        funds.recipient = recoveredOrder.receiver;
+        funds.toInternalBalance = order.useInternalBuyTokenBalance;
+
+        int256[] memory limits = new int256[](tokens.length);
+        // NOTE: Array allocation initializes elements to 0, so we only need to
+        // set the limits we care about.
+        limits[trade.sellTokenIndex] = order.sellAmount.toInt256();
+        limits[trade.buyTokenIndex] = -order.buyAmount.toInt256();
```

Comment:
Does the balance logic allow you to theoretically set two positive numbers here so that you only spend money without receiving anything?

Reply:
I'm not sure I understand.

Reply:
What would happen if all limits were positive integers?

Reply:
The following check is how limits are handled:

```solidity
            require(delta <= limits[i], "SWAP_LIMIT");
```

Where
> Process token deltas, by either transferring tokens from the sender (for positive deltas) or to the recipient (for negative deltas).

So, I imagine the swap would succeed as long as the total spent for any token does not exceed its limit. But in theory, this allows you to, say swap X token A for 0 token B and Y token B for 0 token A, essentially spending a bunch and getting nothing back.

Reply:
Using a negative limit -X for token A enforces that we receive at least X of A in total (aggregating all in and out amounts of token A).